repose-aggregator/components/filters/openapi-validator-filter/src/main/scala/org/openrepose/filters/openapivalidator/HttpServletOAIRequest.scala

```diff
@@ -0,0 +1,150 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.filters.openapivalidator
+
+import java.io.InputStream
+import java.nio.charset.{IllegalCharsetNameException, StandardCharsets, UnsupportedCharsetException}
+import java.util
+import java.util.{Collections, Optional}
+
+import com.atlassian.oai.validator.model.Request
+import com.typesafe.scalalogging.slf4j.StrictLogging
+import javax.servlet.http.HttpServletRequest
+import org.openrepose.commons.utils.http.normal.{QueryParameter, QueryParameterCollection}
+import org.openrepose.filters.openapivalidator.HttpServletOAIRequest._
+
+import scala.collection.JavaConverters._
+import scala.io.Source
+
+/**
+  * A wrapper that exposes an underlying [[HttpServletRequest]] as a [[Request]]
+  * usable by an [[com.atlassian.oai.validator.OpenApiInteractionValidator]].
+  */
+class HttpServletOAIRequest(httpServletRequest: HttpServletRequest)
+  extends Request with StrictLogging {
+
+  private lazy val encoding: String = {
+    Option(httpServletRequest.getCharacterEncoding)
+      .getOrElse(StandardCharsets.ISO_8859_1.name)
+  }
+  private lazy val body: String = {
+    inputStreamToString(httpServletRequest.getInputStream, encoding)
+  }
+
+  // In accordance with OpenAPI specification, only query string parameters (not form data parameters)
+  // are made available:
+  // https://github.com/OAI/OpenAPI-Specification/blob/3.0.2/versions/3.0.2.md#parameter-locations
+  // For an example on validating form data parameters, see:
+  // https://github.com/OAI/OpenAPI-Specification/blob/3.0.2/versions/3.0.2.md#operation-object-example
+  private lazy val queryParameters: Seq[QueryParameter] = {
+    Option(httpServletRequest.getQueryString)
+      .map(new QueryParameterCollection(_))
+      .map(_.getParameters.asScala)
+      .getOrElse(Seq.empty)
+  }
+
+  // Header values are not split on commas.
+  // Splitting is an option when using the `SimpleRequest.Builder`, however doing so many not preserve the semantics
```

Comment:
Once we add RBAC, we will be the ones defining a new validator, and we can split roles in the validation logic for our validator. In other words, we can impose additional constraints on our own data when we process that data rather than forcing all data to conform to our constraints.

Put a different way, splitting header values on commas may change the meaning of the header value. Especially considering that JSON content can be specified in headers (by OpenAPI spec). If we were to split JSON content in a header on commas, we would almost always break it, and schema validation on that header would almost always fail.

All of that being said, upon closer inspection, I suspect that not splitting headers may also cause an issue. Specifically, the validation library does not split headers on validation, which implies that the validation library expects headers to be presented by the request already split. That is probably because the OpenAPI spec says that header parameters are in the `simple` style, which means that multiple values are serialized as being comma-separated.

Upon even closer inspection, it seems like the HTTP specification does actually say that multiple header values can be combined with commas (and thus, can be separated by commas) with a few known exceptions.

See:
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#parameter-object
https://tools.ietf.org/html/rfc7230#section-3.2.2

Reply:
I just realized I wasn't very clear about the result of all of that text.

I will go ahead and split headers on commas. That is the default behavior of the validation library and seems to be the behavior dictated by specification.

Reply:
In that case, I am debating changing this from a wrapper to just a set of functions to create a `SimpleRequest` from an `HttpServletRequest`. Leveraging the `SimpleRequest.Builder` would allow me to use the small amount of logic built into it, plus that approach would almost certainly be less verbose. However, if we did want/need to customize the translation further, we would probably end up writing something like this wrapper again.